Tests/NIOTests/ChannelTests.swift

```diff
@@ -1717,4 +1717,386 @@ public class ChannelTests: XCTestCase {
             XCTFail("unexpected error: \(error)")
         }
     }
+
+    func testNotificationOrder() throws {
+        let group = MultiThreadedEventLoopGroup(numThreads: 1)
+        defer {
+            XCTAssertNoThrow(try group.syncShutdownGracefully())
+        }
+
+        // Handler that verifies the notification order of the different promises + the inbound events that are fired.
+        class SocketChannelLifecycleVerificationHandler: ChannelDuplexHandler {
+            typealias InboundIn = Any
+            typealias OutboundIn = Any
+
+            private var registerPromise: EventLoopPromise<Void>?
+            private var connectPromise: EventLoopPromise<Void>?
+            private var closePromise: EventLoopPromise<Void>?
+
+            public func channelActive(ctx: ChannelHandlerContext) {
+                XCTAssertTrue(ctx.channel.isActive)
+
+                if let registerPromise = self.registerPromise {
+                    XCTAssertTrue(registerPromise.futureResult.fulfilled, "registerPromise not fullfilled before channelActive was called")
+                } else {
+                    XCTFail("register(...) not called before channelActive")
+                }
+
+                if let connectPromise = self.connectPromise {
+                    XCTAssertTrue(connectPromise.futureResult.fulfilled, "connectPromise not fullfilled before channelActive was called")
+                } else {
+                    XCTFail("connect(...) not called before channelActive")
+                }
+
+                XCTAssertNil(self.closePromise)
+                XCTAssertFalse(ctx.channel.closeFuture.fulfilled)
+            }
+
+            public func channelInactive(ctx: ChannelHandlerContext) {
+                XCTAssertFalse(ctx.channel.isActive)
+
+                if let registerPromise = self.registerPromise {
+                    XCTAssertTrue(registerPromise.futureResult.fulfilled, "registerPromise not fullfilled before channelInactive was called")
+                } else {
+                    XCTFail("register(...) not called before channelInactive")
+                }
+
+                if let connectPromise = self.connectPromise {
+                    XCTAssertTrue(connectPromise.futureResult.fulfilled, "connectPromise not fullfilled before channelInactive was called")
+                } else {
+                    XCTFail("connect(...) not called before channelInactive")
+                }
+
+                if let closePromise = self.closePromise {
+                    XCTAssertTrue(closePromise.futureResult.fulfilled, "closePromise not fullfilled before channelInactive was called")
+                } else {
+                    XCTFail("close(...) not called before channelInactive")
+                }
```

Comment:
We can reduce this duplicate code by writing a pair of helper functions:

```swift
func assertFulfilled(promise: EventLoopPromise<Void>?, promiseName: String, trigger: String, setter: String, file: String = #file, line: Uint = #line) {
        if let promise = promise {
            XCTAssertTrue(promise.futureResult.fulfilled, "\(promiseName) not fulfilled before \(trigger) was called", file: file, line: line)
        } else {
            XCTFail("\(setter) not called before \(trigger)", file: file, line: line
        }
    }
```

This would let us truncate this file substantially and better express the criteria of these assertions.